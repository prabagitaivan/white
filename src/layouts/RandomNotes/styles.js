import { makeStyles } from '@material-ui/styles'

const layout = 'RandomNotes'

export const loading = makeStyles(
  {
    root: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      marginTop: '40vh'
    },
    linearProgress: {
      height: 10,
      width: '60%',

      '& .MuiLinearProgress-barColorSecondary': {
        backgroundColor: '#b1b1b1'
      }
    }
  },
  { name: `${layout}Loading` }
)

export const empty = makeStyles(
  {
    root: {
      textAlign: 'center',
      marginTop: '40vh'
    }
  },
  { name: `${layout}Empty` }
)

export const notes = makeStyles(
  {
    img: {
      cursor: 'pointer'
    },
    gridListTileBar: {
      height: 50,

      '& .MuiGridListTileBar-titleWrap': {
        marginLeft: 10,
        marginRight: 10
      },
      '& .MuiGridListTileBar-rootSubtitle': {
        marginLeft: 10,
        marginRight: 10
      },
      '& .MuiGridListTileBar-title': {
        fontSize: 15,
        lineHeight: 'normal',
        color: 'rgba(255, 255, 255, 0.85)'
      },
      '& .MuiGridListTileBar-subtitle': {
        fontSize: 12,
        lineHeight: 'normal',
        color: 'rgba(255, 255, 255, 0.75)'
      }
    }
  },
  { name: `${layout}Notes` }
)
