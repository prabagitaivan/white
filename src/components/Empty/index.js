import React, { memo } from 'react'
import { makeStyles } from '@material-ui/styles'

const useStyles = makeStyles(
  {
    root: {
      textAlign: 'center',
      marginTop: '40vh'
    }
  },
  { name: 'Empty' }
)

export default memo(() => {
  const classes = useStyles()

  return (
    <div className={classes.root}>
      <div>Still Empty</div>
      <span role='img' aria-label='three-wise-monkeys'>
        ğŸ™ˆğŸ™‰ğŸ™Š
      </span>
    </div>
  )
})
