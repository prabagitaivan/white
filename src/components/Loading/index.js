import React, { memo } from 'react'
import { LinearProgress } from '@material-ui/core'
import { makeStyles } from '@material-ui/styles'

const useStyles = makeStyles(
  {
    root: {
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      marginTop: '40vh'
    },
    linearProgress: {
      height: 10,
      width: '60%',

      '& .MuiLinearProgress-barColorSecondary': {
        backgroundColor: '#b1b1b1'
      }
    }
  },
  { name: 'Loading' }
)

export default memo(() => {
  const classes = useStyles()

  return (
    <div className={classes.root}>
      <div>Loading</div>
      <LinearProgress color='secondary' className={classes.linearProgress} />
    </div>
  )
})
