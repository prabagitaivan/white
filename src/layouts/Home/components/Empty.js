import React, { PureComponent } from 'react'
import { empty as styles } from '../styles'

export default class Empty extends PureComponent {
  render () {
    return (
      <div style={styles.root}>
        <div>Still Empty</div>
        <span role='img' aria-label='three-wise-monkeys'>
          ðŸ™ˆðŸ™‰ðŸ™Š
        </span>
      </div>
    )
  }
}
