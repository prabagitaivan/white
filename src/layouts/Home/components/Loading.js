import React, { PureComponent } from 'react'
import { LinearProgress } from '@material-ui/core'
import { loading as styles } from '../styles'

export default class Loading extends PureComponent {
  render () {
    return (
      <div style={styles.root}>
        <div>Loading</div>
        <LinearProgress color='secondary' style={styles.linearProgress} />
      </div>
    )
  }
}
