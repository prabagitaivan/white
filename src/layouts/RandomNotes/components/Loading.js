import React, { memo } from 'react'
import { LinearProgress } from '@material-ui/core'
import { loading as useStyles } from '../styles'

export default memo(() => {
  const classes = useStyles()

  return (
    <div className={classes.root}>
      <div>Loading</div>
      <LinearProgress color='secondary' className={classes.linearProgress} />
    </div>
  )
})
