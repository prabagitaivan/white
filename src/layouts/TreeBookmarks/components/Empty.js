import React, { memo } from 'react'
import { empty as useStyles } from '../styles'

export default memo(() => {
  const classes = useStyles()

  return (
    <div className={classes.root}>
      <div>Still Empty</div>
      <span role='img' aria-label='three-wise-monkeys'>
        ☁️🌧️⛈️
      </span>
    </div>
  )
})
