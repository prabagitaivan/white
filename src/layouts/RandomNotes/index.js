import React, { memo, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { ImageList, ImageListItem, ImageListItemBar } from '@material-ui/core'
import { makeStyles } from '@material-ui/styles'
import Navigator from '../../components/Navigator'
import Content from '../../components/Content'
import { request } from '../../reducers/randomNotes'
import { setPage } from '../../reducers/status'
import Loading from '../../components/Loading'
import Empty from '../../components/Empty'

const useStyles = makeStyles(
  {
    imageListItem: {
      '& .MuiImageListItem-item': {
        borderRadius: 5,
        boxShadow: '0px 5px 5px rgb(0 0 0 / 25%)'
      }
    },
    img: {
      cursor: 'pointer'
    },
    imageListItemBar: {
      height: 50,

      '& .MuiImageListItemBar-titleWrap': {
        marginLeft: 10,
        marginRight: 10
      },
      '& .MuiImageListItemBar-rootSubtitle': {
        marginLeft: 10,
        marginRight: 10
      },
      '& .MuiImageListItemBar-title': {
        fontSize: 15,
        lineHeight: 'normal',
        color: 'rgba(255, 255, 255, 0.85)'
      },
      '& .MuiImageListItemBar-subtitle': {
        fontSize: 12,
        lineHeight: 'normal',
        color: 'rgba(255, 255, 255, 0.75)'
      }
    }
  },
  { name: 'RandomNotes' }
)

export default memo(() => {
  const { desktop } = useSelector(state => state.status)
  const { data, requesting } = useSelector(state => state.randomNotes)
  const dispatch = useDispatch()
  const classes = useStyles()

  const openNote = url => window.open(url)

  useEffect(() => {
    dispatch(setPage('RandomNotes'))
    if (data.length === 0) dispatch(request())
  }, [dispatch])

  return (
    <div>
      <Navigator />
      <Content>
        {requesting ? (
          <Loading />
        ) : data.length === 0 ? (
          <Empty />
        ) : (
          <ImageList cols={desktop ? 4 : 1} rowHeight={250} gap={20}>
            {data
              .filter(note => note.active)
              .reverse()
              .map((note, index) => (
                <ImageListItem key={index} className={classes.imageListItem}>
                  <img
                    src={note.image}
                    alt={note.title}
                    className={classes.img}
                    onClick={() => openNote(note.url)}
                    loading='lazy'
                  />
                  <ImageListItemBar
                    title={note.title}
                    subtitle={note.author}
                    className={classes.imageListItemBar}
                  />
                </ImageListItem>
              ))}
          </ImageList>
        )}
      </Content>
    </div>
  )
})
