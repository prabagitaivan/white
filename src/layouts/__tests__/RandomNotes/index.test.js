import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from '../../../stores'
import Styles from '../../../styles'
import { request } from '../../../reducers/randomNotes'
import { setPage } from '../../../reducers/status'
import Loading from '../../../components/Loading'
import Empty from '../../../components/Empty'
import RandomNotes from '../../RandomNotes'

let store

function renderRandomNotes ({ desktop = true, randomNotes }) {
  const preloadedState = {
    status: { desktop, page: 'RandomNotes' },
    randomNotes
  }
  store = createStore(preloadedState)

  jest.spyOn(store, 'dispatch').mockReturnValue()

  return render(
    <Provider store={store}>
      <Styles>
        <RandomNotes />
      </Styles>
    </Provider>
  )
}

afterEach(() => {
  store.dispatch.mockRestore()
})

describe('layouts RandomNotes', () => {
  describe('snapshots', () => {
    it('contain navigator', () => {
      const { container } = renderRandomNotes({})
      const Navigator = container.querySelector('#navigator')
      expect(Navigator).toBeInTheDocument()
    })
    it('contain loading content when requesting', () => {
      const { container } = renderRandomNotes({})
      const LoadingRender = render(
        <Provider store={store}>
          <Styles>
            <Loading />
          </Styles>
        </Provider>
      )
      const Content = container.querySelector('#content')
      const LoadingContent = LoadingRender.container
      expect(Content.firstChild).toEqual(LoadingContent.firstChild)
    })
    it('contain empty content when requesting result is empty', () => {
      const { container } = renderRandomNotes({
        randomNotes: { requesting: false, data: [] }
      })
      const EmptyRender = render(
        <Provider store={store}>
          <Styles>
            <Empty />
          </Styles>
        </Provider>
      )
      const Content = container.querySelector('#content')
      const EmptyContent = EmptyRender.container
      expect(Content.firstChild).toEqual(EmptyContent.firstChild)
    })
    it('contain data content when requesting result is not empty in desktop', () => {
      const { container } = renderRandomNotes({
        randomNotes: {
          requesting: false,
          data: [
            {
              image: 'image',
              title: 'title',
              url: 'url',
              active: true
            }
          ]
        }
      })
      const Content = container.querySelector('#content')

      expect(Content).toBeInTheDocument()
      expect(Content.firstChild).toMatchSnapshot()
    })
    it('contain data content when requesting result is not empty in mobile', () => {
      const { container } = renderRandomNotes({
        desktop: false,
        randomNotes: {
          requesting: false,
          data: [
            {
              image: 'image',
              title: 'title',
              url: 'url',
              active: true
            }
          ]
        }
      })
      const Content = container.querySelector('#content')

      expect(Content).toBeInTheDocument()
      expect(Content.firstChild).toMatchSnapshot()
    })
    it('show active only in data content', () => {
      const { queryByAltText } = renderRandomNotes({
        randomNotes: {
          requesting: false,
          data: [
            {
              image: 'image1',
              title: 'title1',
              url: 'url1',
              active: false
            },
            {
              image: 'image2',
              title: 'title2',
              url: 'url2',
              active: true
            }
          ]
        }
      })

      expect(queryByAltText('title1')).not.toBeInTheDocument()
      expect(queryByAltText('title2')).toBeInTheDocument()
    })
    it('show subtitle when author provided in data content', () => {
      const { queryByText } = renderRandomNotes({
        randomNotes: {
          requesting: false,
          data: [
            {
              image: 'image1',
              title: 'title1',
              url: 'url1',
              active: true
            },
            {
              image: 'image2',
              title: 'title2',
              url: 'url2',
              author: 'author2',
              active: true
            }
          ]
        }
      })

      expect(queryByText('author1')).not.toBeInTheDocument()
      expect(queryByText('author2')).toBeInTheDocument()
    })
  })
  describe('mounting', () => {
    beforeEach(() => {
      renderRandomNotes({})
    })

    it('set page to random notes after mount', () => {
      expect(store.dispatch).toHaveBeenCalledWith(setPage('RandomNotes'))
    })
    it('request for random notes data after mount', () => {
      expect(store.dispatch).toHaveBeenCalledWith(request())
    })
  })
  describe('userEvent', () => {
    beforeEach(() => {
      jest.spyOn(window, 'open').mockReturnValue()
    })
    afterEach(() => {
      window.open.mockRestore()
    })

    it("open url when click the note's images", () => {
      const { getByAltText } = renderRandomNotes({
        randomNotes: {
          requesting: false,
          data: [
            {
              image: 'image1',
              title: 'title1',
              url: 'url1',
              active: true
            },
            {
              image: 'image2',
              title: 'title2',
              url: 'url2',
              active: true
            }
          ]
        }
      })

      const Image1 = getByAltText('title1')
      userEvent.click(Image1)
      expect(window.open).toHaveBeenCalledWith('url1')

      const Image2 = getByAltText('title2')
      userEvent.click(Image2)
      expect(window.open).toHaveBeenCalledWith('url2')
    })
  })
})
