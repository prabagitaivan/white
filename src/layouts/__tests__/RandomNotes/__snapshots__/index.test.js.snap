// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layouts RandomNotes snapshots contain data content when requesting result is not empty in desktop 1`] = `
<ul
  class="MuiGridList-root"
  style="margin: -10px;"
>
  <li
    class="MuiGridListTile-root"
    style="width: 25%; height: 270px; padding: 10px;"
  >
    <div
      class="MuiGridListTile-tile"
    >
      <img
        alt="title"
        class="WhiteRandomNotes-img"
        src="image"
      />
      <div
        class="MuiGridListTileBar-root WhiteRandomNotes-gridListTileBar MuiGridListTileBar-titlePositionBottom"
      >
        <div
          class="MuiGridListTileBar-titleWrap"
        >
          <div
            class="MuiGridListTileBar-title"
          >
            title
          </div>
        </div>
      </div>
    </div>
  </li>
</ul>
`;

exports[`layouts RandomNotes snapshots contain data content when requesting result is not empty in mobile 1`] = `
<ul
  class="MuiGridList-root"
  style="margin: -10px;"
>
  <li
    class="MuiGridListTile-root"
    style="width: 100%; height: 270px; padding: 10px;"
  >
    <div
      class="MuiGridListTile-tile"
    >
      <img
        alt="title"
        class="WhiteRandomNotes-img"
        src="image"
      />
      <div
        class="MuiGridListTileBar-root WhiteRandomNotes-gridListTileBar MuiGridListTileBar-titlePositionBottom"
      >
        <div
          class="MuiGridListTileBar-titleWrap"
        >
          <div
            class="MuiGridListTileBar-title"
          >
            title
          </div>
        </div>
      </div>
    </div>
  </li>
</ul>
`;
